homeassistant/components/postnl/sensor.py

```diff
@@ -44,21 +48,28 @@ def setup_platform(hass, config, add_entities, discovery_info=None):
 
     try:
         api = PostNL_API(username, password)
+        _LOGGER.debug("Connection with PostNL API succeeded")
 
     except UnauthorizedException:
         _LOGGER.exception("Can't connect to the PostNL webservice")
         return
 
-    add_entities([PostNLSensor(api, name)], True)
+    add_entities([PostNLDelivery(api, name)], True)
+    add_entities([PostNLDistribution(api, name)], True)
+    add_entities([PostNLLetter(api, name)], True)
```

Comment:
The two new classes are more or less duplicating the existing code.

Remove `PostNLDistribution` and `PostNLLetter` but add a data object and a list of the available e sensors. E.g., [`openweathermap`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/openweathermap/sensor.py) or [`bitcoin`](https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/bitcoin/sensor.py) sensor.

Reply:
Not sure if I completely understand. The only thing that is common is a part of the init method, the properties and a part of the update method. I could create a data object, store the name etc in it, but that basically means that with the name property, I have to retrieve it from the data property and the creation still happens in the init of the current class, as it's not a common name.

Also the update method is different between both. Perhaps the shipment foreach loop could be in a data class which handles that, I am just wondering if that's enough gain for you.

Or am I thinking something different then what you ment?